### ISSUE TITLE
Suggestion: quniform()'s semantics should be slightly changed

### ISSUE THREAD

**Issue by lebigot** [hyperopt/hyperopt#328]
`quniform(low, high, q)` is a "quantized" uniform distribution with step `q`.

Now, given the variable names, **I would expect the values to be in [`low`, `high`]** (the same expectation is found in issue  #321), but this is not the case.

I would therefore suggest that the formula for `quniform()` (`round(uniform(low, high) / q) * q`) be replaced by the similar but more expected `low + round(uniform(0, high-low) / q ) * q`, which gives results from `low` to `high` (when the step `q` divides [`low`, `high`]). Note that this formula can yield 8 for `quniform(0, 7.9, 1)`, for instance), but this is the price to pay for handling the likely most common case of `q` dividing `high-low` even in the face of floating-point rounding.

PS: In addition, both the original formula and the proposed formula above actually **don't give a uniform distribution**. A correct expression, that both gives results from `low` to `high` and does so uniformly is for example `low + round(uniform(-q/2, high-low+q/2) / q ) * q` (still assuming that the step `q` divides `high-low` up to a floating-point rounding error).

**Comment by lebigot** [comment]
PS: The suggestion above is all the more important that it is not satisfying to obtain the correct interval by doing `quniform("name", 0, high-low, q) + low`, because the reported values will be those of `"name"` and not those of `"name"` +`low`, which is the actual variable fed to the optimized function.

**Comment by XinyiYS** [comment]
it is in fact an issue: for example quniform(10, 210, 50), the range I need is [10,60,110,160], but it can return 0 using the original formula, which is out of the desired range and can lead to errors downsteam

**Comment by sm-hawkfish** [comment]
Just ran into this issue as well with occasional draws of `0` from `hp.quniform("batch_size", 16, 512, 64)` that were crashing our training loop. We have worked around it by ensuring that our feasible space satisfies `low / q > 0.5` but it was definitely a pain to track this down. 

I see that this issue has been open for almost 2 years, is the delay in addressing this due to resource constraints on the part of the `hyperopt` developers or a lack of consensus on how it should be fixed ?

**Comment by lebigot** [comment]
As far as I see, development on hyperopt resumed maybe a year ago. 
That said, people may want to try the more recent and apparently more convenient optuna package.

**Comment by skylogic004** [comment]
I came here to report a bug with quniform, but now I'm wondering if it's related to this issue. I'm trying to get integers from 1 to 8 to be drawn with equal probability using `hp.quniform('foo', 1, 8, 1)` but, upon sampling a bunch of times, I noticed that it doesn't sample uniformly. Is this also due to the way rounding happens internally?

e.g. This code counts the occurrences:
`pd.Series([sample(scope.int(hp.quniform('foo', 1, 8, 1))) for i in range(10000)]).value_counts().sort_index()`
which outputs:

> 1     699
2    1419
3    1454
4    1467
5    1462
6    1413
7    1373
8     713

Notice how 1 and 8 have ~700 occurrences rather than ~1400 like the rest.

As a workaround, I'm considering using `scope.int(hp.uniform('foo', 1, 8+1))`. Is there any drawbacks to this as far as the optimization process goes or is it effectively the same as quniform?
